clippy_lints/src/needless_pass_by_ref_mut.rs

```diff
@@ -0,0 +1,189 @@
+use super::needless_pass_by_value::requires_exact_signature;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::is_self;
+use clippy_utils::source::snippet;
+use if_chain::if_chain;
+use rustc_errors::{Applicability, Diagnostic};
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::HirIdSet;
+use rustc_hir::{Body, FnDecl, HirId, Impl, ItemKind, Mutability, Node, PatKind};
+use rustc_hir_typeck::expr_use_visitor as euv;
+use rustc_infer::infer::TyCtxtInferExt;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::mir::FakeReadCause;
+use rustc_middle::ty::{self, Ty};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::def_id::LocalDefId;
+use rustc_span::Span;
+use rustc_target::spec::abi::Abi;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check if a `&mut` function argument is actually used mutably.
+    ///
+    /// ### Why is this bad?
+    /// Less `mut` means less fights with the borrow checker. It can also lead to more
+    /// opportunities for parallelization.
+    ///
+    /// ### Example
+    /// ```rust
+    /// fn foo(&mut self, y: &mut i32) -> i32 {
+    ///     self.x + *y
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn foo(&self, y: &i32) -> i32 {
+    ///     self.x + *y
+    /// }
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub NEEDLESS_PASS_BY_REF_MUT,
+    suspicious,
+    "default lint description"
+}
+declare_lint_pass!(NeedlessPassByRefMut => [NEEDLESS_PASS_BY_REF_MUT]);
+
+fn should_skip(span: Span, input: rustc_hir::Ty<'_>, ty: Ty<'_>, arg: &rustc_hir::Param<'_>) -> bool {
+    // All spans generated from a proc-macro invocation are the same...
```

Comment:
I think you should use `is_from_proc_macro` instead. This can be used for `(&FnKind, &Body, HirId, Span)`, which is what you get with `check_fn`. This should be the last check done, due to its cost.